### PR TITLE
making mailru driver compatible with chproxy

### PR DESCRIPTION
**Pull request by evilaffliction**
chproxy on github:
https://github.com/Vertamedia/chproxy

1) Amending Ping logic: send version request with credentials due to chproxy is based on user  authentication 
2) Put sql query for CH for get requests into a request query due to chproxy expects sql query to be there for the get requests. Compatible with CH.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,3 +1,3 @@
-module github.com/mailru/go-clickhouse
```
Comment (bgaifullin) [inline go.mod]: please revert this change

File: conn.go [hunk 2]
```diff
@@ -247,7 +246,13 @@ func (c *conn) buildRequest(ctx context.Context, query string, params []driver.V
 		method = http.MethodPost
 	}
 	c.log("query: ", query)
-	req, err := http.NewRequest(method, c.url.String(), strings.NewReader(query))
+
+	var bodyReader io.Reader
```
Comment (bgaifullin) [inline conn.go]: seems like it can be moved to upper. (where the http method is resolved)

File: conn.go [hunk 3]
```diff
@@ -267,6 +272,11 @@ func (c *conn) buildRequest(ctx context.Context, query string, params []driver.V
 			req.URL.RawQuery = reqQuery.Encode()
 		}
 	}
+	if method == http.MethodGet {
+		reqQuery := req.URL.Query()
+		reqQuery.Add("query", query)
+		req.URL.RawQuery = reqQuery.Encode()
```
Comment (bgaifullin) [inline conn.go]: it would be better to call reqQuery.Encode only one time. previous possible call is in line 272

File: conn_go18.go [hunk 4]
```diff
@@ -5,32 +5,27 @@ package clickhouse
...
-	// make request with empty body, response should be "Ok"
-	u := &url.URL{Scheme: c.url.Scheme, User: c.url.User, Host: c.url.Host, Path: "/"}
-	req, err := http.NewRequest(http.MethodGet, u.String(), nil)
+
+	req, err := c.buildRequest(ctx, "select version()", nil, true)
 	if err != nil {
 		return err
 	}
-	respBody, err := c.doRequest(ctx, req)
+	_, err = c.doRequest(ctx, req)
```
Comment (bgaifullin) [inline conn_go18.go]: what will return from chproxy if CH is not available ?
I think need to call Close for response to avoid resources leak.

File: conn_go18.go [hunk 5]
```diff
@@ -5,32 +5,27 @@ package clickhouse
...
 // Ping implements the driver.Pinger
 func (c *conn) Ping(ctx context.Context) error {
 	if c.transport == nil {
 		return driver.ErrBadConn
 	}
-	// make request with empty body, response should be "Ok"
-	u := &url.URL{Scheme: c.url.Scheme, User: c.url.User, Host: c.url.Host, Path: "/"}
-	req, err := http.NewRequest(http.MethodGet, u.String(), nil)
+
+	req, err := c.buildRequest(ctx, "select version()", nil, true)
```
Comment (bgaifullin) [inline conn_go18.go]: select 1 will not be enough ?

File: go.mod (hunk 1 above)
Comment (evilaffliction) [reply go.mod]: ouch, it wasn't supposed to be here, apologies

File: conn_go18.go (hunk 5 above)
Comment (evilaffliction) [reply conn_go18.go]: I thought about it, may be yes, it'll be even better since we can scan a body of a response and check it to be exactly "1".

**Comment by DoubleDi** [comment]
@evilaffliction will you make the fixes, so we can merge your PR?

**Comment by evilaffliction** [comment]
yes, can you please give me 2 days more? everything will be completed by
tomorrow.
My apologies, was extremely overloaded on my job.

вт, 6 авг. 2019 г. в 19:18, Denis <notifications@github.com>:

**Comment by evilaffliction** [comment]
It seems that i've fixed all notations from the review/pull request

вт, 6 авг. 2019 г. в 19:56, Marat Seifullin <seifullinm@gmail.com>:

**Comment by DoubleDi** [comment]
@bgaifullin can we merge this PR now?

**Comment by evilaffliction** [comment]
I'll resolve conflicts this evening

**Comment by DoubleDi** [comment]
@evilaffliction  can you give a status update for this PR?

**Comment by shashankkoppar** [comment]
Guys what is the update on this? :)
